### ISSUE TITLE
LLVM errors at linking stage when building on linux.

### ISSUE THREAD

**Issue by 7oxicshadow** [RPCS3/rpcs3#2227]
I am using clang to build and when it gets to 100% the linker fails with the following error:

```
[100%] Linking CXX executable ../bin/rpcs3
CMakeFiles/rpcs3.dir/Emu/Cell/PPUTranslator.cpp.o:(.rodata._ZTIN4llvm10SelectInstE[_ZTIN4llvm10SelectInstE]+0x10): undefined reference to `typeinfo for llvm::Instruction'
CMakeFiles/rpcs3.dir/Emu/Cell/PPUTranslator.cpp.o:(.rodata._ZTIN4llvm8FCmpInstE[_ZTIN4llvm8FCmpInstE]+0x10): undefined reference to `typeinfo for llvm::CmpInst'
clang-3.8: error: linker command failed with exit code 1 (use -v to see invocation)
rpcs3/CMakeFiles/rpcs3.dir/build.make:10572: recipe for target 'bin/rpcs3' failed
make[2]: *** [bin/rpcs3] Error 1
CMakeFiles/Makefile2:233: recipe for target 'rpcs3/CMakeFiles/rpcs3.dir/all' failed
make[1]: *** [rpcs3/CMakeFiles/rpcs3.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

Does linux use LLVM as I see no mention of it when running cmake?

**Comment by Nekotekina** [comment]
Try to compile PPUTranslator.cpp without RTTI. Example:
https://github.com/RPCS3/rpcs3/blob/master/rpcs3/CMakeLists.txt#L156

**Comment by 7oxicshadow** [comment]
I modified CMakeLists to look like the following:
```
	if (NOT MSVC)
		set_source_files_properties(${RPCS3_SRC_DIR}/../Utilities/JIT.cpp PROPERTIES COMPILE_FLAGS -fno-rtti)
		set_source_files_properties(${RPCS3_SRC_DIR}/Emu/Cell/PPUTranslator.cpp PROPERTIES COMPILE_FLAGS -fno-rtti)
	endif()
```

and it fails part way through compilation with the following error:

```
[ 39%] Building CXX object rpcs3/CMakeFiles/rpcs3.dir/Emu/Cell/PPUTranslator.cpp.o
In file included from /mnt/games/native_emulators/source_files/rpcs3/rpcs3/Emu/Cell/PPUTranslator.cpp:3:
In file included from /mnt/games/native_emulators/source_files/rpcs3/rpcs3/Emu/Cell/PPUTranslator.h:12:
/mnt/games/native_emulators/source_files/rpcs3/rpcs3/../rpcs3/Emu/Cell/PPUAnalyser.h:59:63: error: no type named 'vector' in namespace 'std'
extern void ppu_validate(const std::string& fname, const std::vector<ppu_function>& funcs, u32 reloc);
                                                         ~~~~~^
/mnt/games/native_emulators/source_files/rpcs3/rpcs3/../rpcs3/Emu/Cell/PPUAnalyser.h:59:69: error: expected ')'
extern void ppu_validate(const std::string& fname, const std::vector<ppu_function>& funcs, u32 reloc);
                                                                    ^
/mnt/games/native_emulators/source_files/rpcs3/rpcs3/../rpcs3/Emu/Cell/PPUAnalyser.h:59:25: note: to match this '('
extern void ppu_validate(const std::string& fname, const std::vector<ppu_function>& funcs, u32 reloc);
                        ^
/mnt/games/native_emulators/source_files/rpcs3/rpcs3/../rpcs3/Emu/Cell/PPUAnalyser.h:61:13: error: no type named 'vector' in namespace 'std'
extern std::vector<ppu_function> ppu_analyse(const std::vector<std::pair<u32, u32>>& segs, const std::vector<std::pair<u32, u32>>& secs, u32 lib_toc);
       ~~~~~^
/mnt/games/native_emulators/source_files/rpcs3/rpcs3/../rpcs3/Emu/Cell/PPUAnalyser.h:61:19: error: expected unqualified-id
extern std::vector<ppu_function> ppu_analyse(const std::vector<std::pair<u32, u32>>& segs, const std::vector<std::pair<u32, u32>>& secs, u32 lib_toc);
                  ^
4 errors generated.
rpcs3/CMakeFiles/rpcs3.dir/build.make:3450: recipe for target 'rpcs3/CMakeFiles/rpcs3.dir/Emu/Cell/PPUTranslator.cpp.o' failed
make[2]: *** [rpcs3/CMakeFiles/rpcs3.dir/Emu/Cell/PPUTranslator.cpp.o] Error 1
CMakeFiles/Makefile2:233: recipe for target 'rpcs3/CMakeFiles/rpcs3.dir/all' failed
make[1]: *** [rpcs3/CMakeFiles/rpcs3.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by Nekotekina** [comment]
Thanks. I remembered now, it seems troublesome. What is the version of LLVM, btw?

**Comment by 7oxicshadow** [comment]
llvm-devel-3.8.0-1.fc24.x86_64 is installed on my machine (Fedora) but I also have the latest LLVM submodule folder which was pulled in using:

`git submodule update --init --recursive`

I am not sure which it is using as the cmake command does not make any reference to which LLVM directory it has found?

**Comment by Nekotekina** [comment]
I can suggest trying different compilers, because it's not very easy to fix properly.

**Comment by Nekotekina** [comment]
Oops, try to remove LLVM submodule.

**Comment by 7oxicshadow** [comment]
I have tried removing the LLVM submodule and it made no difference.

I have switched to the gcc compiler (6.2.1) and it fails in the same way with and without -fno-rtti.

Doesnt look like this will be an easy issue to resolve for a while. Is there a way of forcing a build without LLVM until this issue is resolved or is it a key dependency?

**Comment by Nezarn** [comment]
@7oxicshadow which linux do you use? it should work on Ubuntu (and its variants like Lubuntu), and on Arch. (at least i tried it on these distros and both worked fine)

**Comment by 7oxicshadow** [comment]
I use fedora 24. I built it using clang about a month ago without any problems but the latest git revision has this problem.

Have you built the latest git on both ubuntu and arch?

When i get chance i will try going back a few versions to see if it make a difference.
Also do you knkw what version of clang you are building with?

**Comment by Nezarn** [comment]
@7oxicshadow if it worked a month ago, it still should work, since last commit was nearly a month ago, and the last time someone touched rpcs3 in LLVM (or anything core related, since these days we just seen only renderer fixes) was a long time ago, it still should work.

On arch i have clang 3.8.1 with LLVM 3.8.1-1 (and im sure ubuntu has older clang and llvm, but i will need to reinstall ubuntu in VM)

If you can wait for some time (im busy now) i can even check fedora in VM

**Comment by 7oxicshadow** [comment]
If you get chance to try it in a VM (64bit) it would be great so I can compare. If you can't it isn't a massive problem, I will keep trying a few things to see if I can find the cause for it.

If it isn't a result of source code problems then I can only assume that there has been a compiler update (or llvm update) that has introduced the problem....

**Comment by 7oxicshadow** [comment]
oops, closed by accident. sorry

**Comment by Nezarn** [comment]
@7oxicshadow After installing Fedora 24 and updated, installed dependencies, and manually making git-version.h (for some reason this never worked for me in linux)

![vmware_2016-11-17_18-47-39](https://cloud.githubusercontent.com/assets/3817985/20400767/5c34921a-acf6-11e6-9221-81d37859eca6.png)

As you can see it works fine. (so something is broken on your system)
Also installing the devel package for LLVM isn't enough, you need to install llvm-static-3.8.0-1.fc24.x86_64 too

**Comment by 7oxicshadow** [comment]
@Nezarm Thanks for taking the time to build on a virtual machine. I have installed llvm-static but unfortunately it did not have seem to have any effect on this problem.

I will try building using GCC again shortly to see if it has any effect. Other than that I will wait until next week when Fedora 25 is released and will do a clean OS install to see if it fixes things.

If you manage to think of any other ideas to try, drop them in here and I will give it a go.

**Comment by Nezarn** [comment]
@7oxicshadow try running  
>make clear

And also clean cmake generated files and run cmake again (and make sure you're running it from the main directory, and not from rpcs3/rpcs3)

Also double check if git checkout every submodule (sometimes it just makes empty directories), and as a last resort you should clone it again and just use git submodule update --init

**Comment by kirbyfan64** [comment]
@Nezarn `make clean`?

**Comment by Nezarn** [comment]
@kirbyfan64 typo :3 (fix'd)

**Comment by 7oxicshadow** [comment]
Ok so I finally fixed the issue. What a nightmare. I believe @Nezarn was correct in saying that my machine was (or still is) broken.

My first test was to remove all llvm packages from my machine and then try to rebuild rpcs3 from a clean build. The idea being that the cmake command should fail saying that there is missing dependencies. Strangely it didnt fail and carried on building as if it had found something??????

I did a "dnf info installed" command to see installed packages and it listed some 32bit llvm libraries but more importantly is listed over 100 entries installed via koji-override-0. I compared this against a freshly installed version of Fedora on a VM and it has no mention of koji-override-0.

I then did a "locate *llvm" command and it found a number of folders that didnt exist? I had to run a "updatedb" command to fix the references.

I then manually removed all remaining llvm folders that I believed should no longer exist.

I then did another clean build of rpcs3 and lo and behold the cmake command failed complaining that it cannot locate LLVM.

I reinstalled the following dependencies "llvm llvm-static llvm-devel" and started the build again. Everything built successfully and the program now appears to be functioning as expected. I am definitely doing a fresh install of fedora 25 when its released. I think my build is completely messed up somehow.  

Many thanks to all who helped. It took a while but I got there in the end.

Thanks.
